File: Tree/872_LeafSimilarTree.cs
// 872. Leaf-Similar Trees.
// Tags: tree, DFS 
// 

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution 
{
    public bool LeafSimilar(TreeNode root1, TreeNode root2) 
    {
        // Edge case.
        if(root1 == null || root2 == null)
            return false;

        List<int> leaves1 = new();
        List<int> leaves2 = new();

        _DFS(root1, ref leaves1);
        _DFS(root2, ref leaves2);

        if(leaves1.Count != leaves2.Count)
            return false;
        else 
        {
            for(int i = 0; i < leaves1.Count; i++)
            {
                if(leaves1[i] != leaves2[i])
                    return false;
            }
        }
        return true;
    }

    private void _DFS(TreeNode node, ref List<int> leaves)
    {
        if(node == null)
            return;

        else if(node.left == null && node.right == null)
            leaves.Add(node.val);

        _DFS(node.left, ref leaves);
        _DFS(node.right, ref leaves);
    }
}

File: Tree/872_LeafSimilarTree.rs
// 872. Leaf-Similar Trees.
// Tags: tree, DFS
//

// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;

type OptNode = Option<Rc<RefCell<TreeNode>>>;

impl Solution {
    pub fn leaf_similar(root1: OptNode, root2: OptNode) -> bool {

        // Edge cases.
        if let None = root1 {
            return false;
        }
        if let None = root2 {
            return false;
        }

        let (mut leaves1, mut leaves2) = (Vec::new(), Vec::new());
        Solution::dfs(&root1, &mut leaves1);
        Solution::dfs(&root2, &mut leaves2);

        leaves1 == leaves2
    }

    fn dfs(node: &OptNode, leaves: &mut Vec<i32>) {
        if let Some(curr) = node {
            let curr = curr.borrow();

            // Leaf case.
            if curr.left.is_none() && curr.right.is_none() {
                leaves.push(curr.val);
                return;
            }

            Solution::dfs(&curr.left, leaves);
            Solution::dfs(&curr.right, leaves);
        }
    }
}